File: 0020_valid_parentheses/is_valid.java
class Solution {
    // Time: O(n)
    // Space: O(n)
    public boolean isValid(String s) {
        Stack<Character> stack = new Stack<Character>();
        HashMap<Character, Character> hashMap = new HashMap<Character, Character>() {
            {
                put('(', ')');
                put('[', ']');
                put('{', '}');
            }  
        };
        for (char c: s.toCharArray()) {
            if (hashMap.containsKey(c)) {
                stack.add(c);
                continue;
            }
            if (stack.isEmpty())
                return false;
            char tmp = stack.pop();
            char compliment = hashMap.get(tmp);
            if (c != compliment)
                return false;
        }
        return stack.isEmpty();
    }
}

File: 0020_valid_parentheses/is_valid.js
// Time:    O(n)
// Space:   O(n)
const isValid = (str) => {
    const map = {
        ']': '[',
        ')': '(',
        '}': '{',
    }
    const stack = []
    for (let c of str) {
        if (c in map) {
            if (stack.pop() != map[c]) {
                return false
            }
        } else stack.push(c)
    }
    return stack.length === 0
}
